KOGITO-3759: support MongoDB persistence with transaction and strong consistency with MongoDB and Debezium (#3925)
kiegroup/kie-docs: doc-content/kogito-docs/src/main/asciidoc/configuration/chap-kogito-configuring.adoc

* https://github.com/kiegroup/kogito-examples/tree/stable/process-kafka-quickstart-springboot[`process-kafka-quickstart-springboot`]: Example on Spring Boot
// end::proc-messaging-enabling[]

[id="proc-events-mongodb-debezium_{context}"]
=== Enabling consistency in process data and {PRODUCT} events using MongoDB and Debezium

[role="_abstract"]

> To avoid the data inconsistencies in process services, the {PRODUCT} services must update the process data and publish events within a unit of work. However, Apache Kafka does not support distributed transactions. Therefore, in order to resolve this problem, you can use an outbox table in the database of the {PRODUCT} service.

Author: ```suggestion
To avoid the data inconsistencies in process services, the {PRODUCT} services must update the process data and publish events within a unit of work atomically. However, Apache Kafka does not support distributed transactions. In order to resolve this problem, you can apply the outbox pattern with {PRODUCT} service.
```
The solution to this problem is to use `transaction`, which is a database feature to ensure multiple operations are done atomically.
Usually, events are directly published to Kafka, they are not stored to db. The solution is to create outbox tables to store the events to db, instead of publishing to Kafka directly. 
For any single unit of work, Kogito performs all the db actions in one transaction, which means storing process data and events to db in one transaction.
This solution is called `outbox pattern`.
So for the last sentence, I think we should either mention `transaction` or use `outbox pattern` instead of `outbox table in the database`.

> When you store the data of a process service in a unit of work, you can also add the records that represent the events to be sent into the outbox table related to the same transaction. After the transaction, a separate message relay process publishes the events that are added into the outbox table to Kafka broker asynchronously. For more information about the outbox table, see https://debezium.io/blog/2019/02/19/reliable-microservices-data-exchange-with-the-outbox-pattern/[Debezium blog].

Author: ```suggestion
When you store the data of a process service in a unit of work, you can also add the records that represent the events to be sent into the outbox tables in the same transaction. After that, a separate message relay process publishes the events that are added into the outbox tables to Kafka broker asynchronously. For more information about the outbox pattern, see https://debezium.io/blog/2019/02/19/reliable-microservices-data-exchange-with-the-outbox-pattern/[Debezium blog].
```
There can be multiple outbox tables (i.e. MongoDB collections) for different types of events.

Reviewer: Done

Reviewer: Done

Reviewer: Done

https://debezium.io/[Debezium] is useful for message relay, which comes with `Change Data Capture` connectors for various databases, such as MongoDB. It captures new entries in MongoDB collections and stream those entries to Apache Kafka.

> {PRODUCT} enables you to store the data of a process and events of the same unit in a single transaction using MongoDB.

Author: ```suggestion
{PRODUCT} enables you to store the data of a process and events of the same unit of work in a single transaction using MongoDB.
```

> For this procedure, a {PRODUCT] service is configured to store the process data and events in a single transaction using MongoDB collection. Also, a Debezium MongoDB connector is configured and started to stream the events from MongoDB collection to Apache Kafka topics.

Author: ```suggestion
For this procedure, a {PRODUCT} service is configured to store the process data and events in a single transaction using MongoDB. Also, a Debezium MongoDB connector is configured and started to stream the events from MongoDB collections to Apache Kafka topics.
```

Reviewer: Done

    "key.converter.schemas.enable": "false",
    "value.converter": "org.apache.kafka.connect.json.JsonConverter",
    "value.converter.schemas.enable": "false",
    "collection.include.list": "kogito.kogitoprocessinstancesevents,kogito.kogitousertaskinstancesevents,kogito.kogitovariablesevents",
    "transforms": "unwrap,reroute",
    "transforms.unwrap.type": "io.debezium.connector.mongodb.transforms.ExtractNewDocumentState",
    "transforms.unwrap.array.encoding": "array",
    "transforms.unwrap.drop.tombstones": "false",
    "transforms.unwrap.delete.handling.mode": "drop",
    "transforms.unwrap.operation.header": "false",
    "transforms.reroute.type": "io.debezium.transforms.ByLogicalTableRouter",
    "transforms.reroute.topic.regex": "(.*)kogito(.*)events(.*)",
    "transforms.reroute.topic.replacement": "kogito-$2-events",
    "transforms.reroute.key.enforce.uniqueness": "false",
    "skipped.operations": "u,d",
    "tombstones.on.delete": "false"
  }
}
----

In the previous example configuration, replace the server host, port, replica set, database, and collections information according to your MongoDB, Kafka installation, and {PRODUCT} project configuration.
--

> You can also use the following example application for consistency in a {PRODUCT} service using MongoDB and Debezium:

Reviewer: Done

* https://www.mongodb.com/[MongoDB] version 4.4 or later is installed, started, and `oplog` is enabled. For information about MongoDB setting up for Debezium, see the https://debezium.io/documentation/reference/connectors/mongodb.html#setting-up-mongodb[MongoDB connector documentation].
* https://debezium.io/[Debezium] version 1.7 or later is installed and started. For information about Debezium installation and configuration, see the https://debezium.io/documentation/reference/[Debezium documentation].

.Procedure
. Add the following dependencies to the `pom.xml` file of your {PRODUCT} project to store {PRODUCT} events to MongoDB:
+
--
.On Quarkus
[source, xml]
----
<dependency>
    <groupId>org.kie.kogito</groupId>
    <artifactId>kogito-addons-quarkus-events-mongodb</artifactId>
</dependency>
----

.On Spring Boot
[source,xml]
----
<dependency>
    <groupId>org.kie.kogito</groupId>
    <artifactId>kogito-addons-springboot-events-mongodb</artifactId>
</dependency>
----
--

> . Configure the MongoDB database and collections to store the {PRODUCT} events to the MongoDB collections, such as outbox tables:

Author: ```suggestion
. Configure the MongoDB database and collections for {PRODUCT} services to store the {PRODUCT} events to the MongoDB collections:
```
